block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -62,6 +64,90 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
 
     def invalidBlocks: F[Set[BlockMetadata]] = invalidBlocksSet.pure[F]
 
+    /**
+      * Truncate full DAG according to some view, defined by latest messages.
+      */
+    override def truncate(
+        targetLatestMessages: Map[Validator, BlockHash],
+        findLfb: Map[Validator, BlockHash] => F[BlockHash]
+    ): F[BlockDagRepresentation[F]] = {
+
+      val lmAll       = this.latestMessagesMap.values.toSet
+      val seenSenders = targetLatestMessages.keySet
+
+      for {
+        lfb         <- findLfb(targetLatestMessages)
+        lmFinalized <- this.latestFinalized(lfb, seenSenders).map(_.valuesIterator.toSet)
+        lmSeen      <- targetLatestMessages.values.toList.traverse(this.lookupUnsafe).map(_.toSet)
+
+        // for all known latest messages, collect all self justifications until first finalized message found
+        toAdjust = {
+          val seenSeqNums = lmSeen.map(m => (m.sender, m.seqNum)).toMap
+          val finSeqNums  = lmFinalized.map(m => (m.sender, m.seqNum)).toMap
+          val seen        = (m: BlockMetadata) => seenSeqNums.get(m.sender).exists(_ >= m.seqNum)
+          val finalized   = (m: BlockMetadata) => finSeqNums.get(m.sender).exists(_ >= m.seqNum)
+
+          Stream
+            .unfoldLoopEval(lmAll.toVector) { messages =>
+              for {
+                metas                    <- messages.traverse(this.lookupUnsafe).map(_.filterNot(finalized))
+                (toUnfinalize, toRemove) = metas.partition(seen)
+
+                parents = metas.flatMap(_.parents).distinct
+                next    = parents.nonEmpty.guard[Option].as(parents)
+                out     = toRemove.map((_, true)) ++ toUnfinalize.map((_, false))
+              } yield (Stream.emits(out.toList), next)
+            }
+            .flatten
+            .fold((Map.empty[Long, Set[BlockHash]], Map.empty[Long, Set[BlockHash]], childMap)) {
+              case ((removeAcc, unfinalizeAcc, childMap), (m, shouldRemove)) =>
+                val height = m.blockNum
+                if (shouldRemove)
+                  (
+                    removeAcc + (height -> (removeAcc.getOrElse(height, Set()) + m.blockHash)),
+                    unfinalizeAcc,
+                    // Remove from children map key for message, adjust keys for parents.
+                    // As some parent might be already removed by previous iterations, use filter.
+                    m.parents.filter(childMap.contains).foldLeft(childMap - m.blockHash) { (a, p) =>
+                      a + (p -> (a(p) - m.blockHash))
+                    }
+                  )
+                else
+                  (
+                    removeAcc,
+                    unfinalizeAcc + (height -> (unfinalizeAcc.getOrElse(height, Set()) + m.blockHash)),
+                    childMap
+                  )
```

Comment:
I think we need to traverse into `children` for each parent to find newer blocks not seen by parent relation.
```
      c2 (a2, b2, c1) <- full dag
a3 b4             <- latest (a3, b4, c1)
   b3
a2 b2   
a1 b1 c1
```
`c2` should not be visible in truncated dag. It has parents (a2, b2, c1) which means child map for a1 and b1 will contain c2.

Reply:
But this piece of code is for unfinalized messages. 
If msg should be removed it is 
```scala
m.parents.filter(childMap.contains).foldLeft(childMap - m.blockHash) { (a, p) =>
                      a + (p -> (a(p) - m.blockHash))
                    }
```
as per several lines above

Reply:
Yes, my comment is for remove case.

My concern is not for keys in the child map but for values which could have children not visible from the target latest messages.
I'm not sure how we will remove `c2` from children of `a1`, `b1` or `c1` in child map?

Reply:
c2 has parents `(a2, b2, c1)`, so for each of them when removing c2 `a + (p -> (a(p) - m.blockHash))` will be invoked? I must miss some point, Im not sure which one...

Reply:
When we traverse from target latest blocks (a3, b4, c1) we will not see c2.

Reply:
But we traverse from the most latest blocks, so from `a3 b4 c2`
https://github.com/rchain/rchain/blob/c093633ccc19bd6a1e486cf07becb301163a27f9/block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala#L90-L92
`val lmAll       = this.latestMessagesMap.values.toSet`

Reply:
OK, that makes sense. Is this means we will rebuild childs for a1, b1 and c1 which has c2 as a child?

Reply:
I'm nit sure - why we need to rebuild? When `c2` is removed - we remove `c2` key from children map (in this particular case there is no such key as no children), and update values for keys `a2,b2,c1`. Same os for every other message - remove key and update keys for parents.

Reply:
I see, we will update childs for a1, b1 and c1 via parents of c2. This is what I mean rebuild.

Reply:
I'm still not sure where is the problem :) if we are removing some hash - it should be removed as a parent, so key completely removed, plus we need to update records for all parents